cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,387 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/ledgerbucketwindow"
+)
+
+const (
+	transactionTableName = "transactions"
+)
+
+type Transaction struct {
+	Result           []byte   // XDR encoded xdr.TransactionResult
+	Meta             []byte   // XDR encoded xdr.TransactionMeta
+	Envelope         []byte   // XDR encoded xdr.TransactionEnvelope
+	Events           [][]byte // XDR encoded xdr.DiagnosticEvent
+	FeeBump          bool
+	ApplicationOrder int32
+	Successful       bool
+	Ledger           ledgerbucketwindow.LedgerInfo
+}
+
+// TransactionWriter is used during ingestion
+type TransactionWriter interface {
+	InsertTransactions(lcm xdr.LedgerCloseMeta) error
+	RegisterMetrics(ingest, insert, count prometheus.Observer)
+}
+
+// TransactionReader is used to serve requests and just returns a cloned,
+// read-only DB session to perform actual DB reads on.
+type TransactionReader interface {
+	NewTx(ctx context.Context) (TransactionReaderTx, error)
+	GetLedgerRange(ctx context.Context) ledgerbucketwindow.LedgerRange
+}
+
+// TransactionReaderTx provides all of the public ways to read from the DB.
+// Note that `Done()` *MUST* be called to clean things up.
+type TransactionReaderTx interface {
+	GetTransaction(log *log.Entry, hash xdr.Hash) (Transaction, bool, ledgerbucketwindow.LedgerRange)
+	GetLedgerRange() ledgerbucketwindow.LedgerRange
+	Done() error
+}
+
+type transactionHandler struct {
+	db         db.SessionInterface
+	stmtCache  *sq.StmtCache
+	passphrase string
+
+	ingestMetric, insertMetric, countMetric prometheus.Observer
+}
+
+type transactionReaderTx struct {
+	ctx        context.Context
+	db         db.SessionInterface
+	passphrase string
+}
+
+func NewTransactionReader(db db.SessionInterface, passphrase string) TransactionReader {
+	return &transactionHandler{db: db, passphrase: passphrase}
+}
+
+func NewTransactionWriter(db db.SessionInterface, stmtCache *sq.StmtCache, passphrase string) TransactionWriter {
+	return &transactionHandler{db: db, stmtCache: stmtCache, passphrase: passphrase}
+}
+
+func (txn *transactionHandler) InsertTransactions(lcm xdr.LedgerCloseMeta) error {
+	start := time.Now()
+	txCount := lcm.CountTransactions()
+	L := log.
+		WithField("ledger_seq", lcm.LedgerSequence()).
+		WithField("tx_count", txCount)
+
+	if txn.stmtCache == nil {
+		return errors.New("TransactionWriter incorrectly initialized without stmtCache")
+	}
+
+	if txCount == 0 {
+		L.Warnf("No transactions present in ledger: %+v", lcm)
```

Comment:
I don't think this is worth logging. If you're running against  stand alone network there will be lots of these cases logged